### ISSUE TITLE
New Redis-py version 5.0.0 breaks on Python 3.7

### ISSUE THREAD

**Issue by rahmed-hartree** [redis/redis-py#2892]
**Version**: redis-py version 5.0.0

**Platform**: Python 3.7.9

**Description**: 

![image](https://github.com/redis/redis-py/assets/135608588/df0417a6-f5e7-45c4-8bde-11d83635282d)

redis-py version 5.0 is meant to be the last working version for python 3.7 but it seems to break due to a type issue. 

```
    from redis.cluster import RedisCluster
  File "/root/.local/lib/python3.7/site-packages/redis/__init__.py", line 3, in <module>
    from redis import asyncio  # noqa
  File "/root/.local/lib/python3.7/site-packages/redis/asyncio/__init__.py", line 1, in <module>
    from redis.asyncio.client import Redis, StrictRedis
  File "/root/.local/lib/python3.7/site-packages/redis/asyncio/client.py", line 673, in <module>
    class PubSub:
  File "/root/.local/lib/python3.7/site-packages/redis/asyncio/client.py", line 744, in PubSub
    def close(self) -> Awaitable[NoReturn]:
  File "/usr/local/lib/python3.7/typing.py", line 251, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.7/typing.py", line 626, in __getitem__
    params = tuple(_type_check(p, msg) for p in params)
  File "/usr/local/lib/python3.7/typing.py", line 626, in <genexpr>
    params = tuple(_type_check(p, msg) for p in params)
  File "/usr/local/lib/python3.7/typing.py", line 135, in _type_check
    raise TypeError(f"Plain {arg} is not valid as type argument")
TypeError: Plain typing.NoReturn is not valid as type argument
```

**Comment by dvora-h** [comment]
@rahmed-hartree Thanks for reporting this. 
What did you try to do that causes this error? I want to reproduce it so I can fix it.

**Comment by rahmed-hartree** [comment]
Was just using redis-py >= 4.2.2, failed when running unit tests in a pipeline when redis-py version 5.0.0 was released a couple of hours ago. 

Just adding `from redis.cluster import RedisCluster` to a py file using vesion 5.0.0 should reproduce the issue

**Comment by dvora-h** [comment]
It works for me, with python 3.7.16, using redis-py 5.0.0.
<img width="1034" alt="Screen Shot 2023-08-15 at 16 33 19" src="https://github.com/redis/redis-py/assets/67596500/1eef7f2d-7ec9-485e-8218-d8c892bbc35e">

**Comment by mszb** [comment]
Having similar issue. Downgraded to 4.6.0 work for me.
```
File "/usr/lib/python3.9/site-packages/django_redis/cache.py", line 31, in _decorator
    return method(self, *args, **kwargs)
  File "/usr/lib/python3.9/site-packages/django_redis/cache.py", line 103, in delete
    result = self.client.delete(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/django_redis/client/default.py", line 354, in delete
    client = self.get_client(write=True)
  File "/usr/lib/python3.9/site-packages/django_redis/client/default.py", line 105, in get_client
    self._clients[index] = self.connect(index)
  File "/usr/lib/python3.9/site-packages/django_redis/client/default.py", line 118, in connect
    return self.connection_factory.connect(self._server[index])
  File "/usr/lib/python3.9/site-packages/django_redis/pool.py", line 71, in connect
    params = self.make_connection_params(url)
  File "/usr/lib/python3.9/site-packages/django_redis/pool.py", line 43, in make_connection_params
    "parser_class": self.get_parser_cls(),
  File "/usr/lib/python3.9/site-packages/django_redis/pool.py", line 100, in get_parser_cls
    return import_string(cls)
  File "/usr/lib/python3.9/site-packages/django/utils/module_loading.py", line 22, in import_string
    raise ImportError('Module "%s" does not define a "%s" attribute/class' % (
ImportError: Module "redis.connection" does not define a "HiredisParser" attribute/class
```

Python version `3.8.17`
```
django-redis==5.0.0
hiredis==2.2.3
redis==5.0.0
```

**Comment by rahmed-hartree** [comment]
@dvora-h 
The pipeline was using python version 3.7.1, error is below:
![image](https://github.com/redis/redis-py/assets/135608588/5591b01b-73f2-4274-9c6e-ebcb7c265d67)

**Comment by chayim** [comment]
Hi folks - can you help provide some more information? In the attachment below you can see two dockers, both running with Python 3.7.16. The top window has hiredis, the bottom window doesn't.

Both dockers were started by running `docker run -it python:3.7.16 bash`. Next:
```pip install redis```

Finally, in the topmost docker `pip install hiredis` was run.
![Screenshot from 2023-08-16 09-24-38](https://github.com/redis/redis-py/assets/1333958/27b04d32-38a2-4b97-82bf-1c5347b8e977)

**Comment by yang2lalang** [comment]
Same issue here

![image](https://github.com/redis/redis-py/assets/12873450/dfda88b6-640e-4d6b-a4c9-799a48654214)

**Comment by Biszu** [comment]
> Having similar issue. Downgraded to 4.6.0 work for me.
> 
> ```
> (...)
>   File "/usr/lib/python3.9/site-packages/django/utils/module_loading.py", line 22, in import_string
>     raise ImportError('Module "%s" does not define a "%s" attribute/class' % (
> ImportError: Module "redis.connection" does not define a "HiredisParser" attribute/class
> ```
> 
> ```
> django-redis==5.0.0
> hiredis==2.2.3
> redis==5.0.0
> ```

I have the exact same problem (`Module "redis.connection" does not define a "HiredisParser"`) with redis 5.0.0 and Python 3.10, so I'm not sure if the original issue is not specific to Python 3.7 or the problem with Hiredis is something else.
In my case downgrade to 4.6.0 helped as well.

**Comment by rahmed-hartree** [comment]
@chayim 

As shown above, the error came from using Python version 3.7.1 which is still Python 3.7, the issue is not with Python version 3.7.16

**Comment by tirkarthi** [comment]
This was a CPython issue fixed in 3.7.2. So version above 3.7.1 will not have this.

Ref : https://bugs.python.org/issue34921

**Comment by dvora-h** [comment]
As this is a python issue that was fixed in python 3.7.2 we recommend to upgrade your python to 3.7.2+.

Regarding the HiredisParser - yes, we moved it and changed it to private, that means that we can now break it without notify about breaking change.
